README.md

```diff
@@ -165,63 +164,64 @@ class GeoEngineer::Resources::AwsElb < GeoEngineer::Resource
 
-```ruby
-class LoadBalancedInstance < Template
-  attr_reader :elb, :elb_sg
-
-  def initialize(name, project, parameters)
-    super(name, project, parameters)
-    # { listeners: [{ in: out: }]}
-    listeners = parameters[:listeners] || []
-
-    # Create the Security Groups
-    elb_sg = resource("aws_security_group", "#{name}_allow_http") {
-      name         "#{name}_elb_sg"
-      description  ""
-      vpc_id       env.vpc_id
-      for l in listeners
-        ingress {
-          from_port    l[:in]
-          to_port      l[:in]
-          protocol     "tcp"
-          cidr_blocks  ["0.0.0.0/0"]
-        }
-      end
-      tags { Name "#{name}_elb_sg" }
-    }
+GPS is a proposed abstraction between developers and GeoEngineer resources. GPS:
 
-    # ELB
-    elb = resource("aws_elb", "main-web-app") {
-      name             "#{name}_elb"
-      security_groups  [elb_sg]
-      subnets          [env.subnet]
-      for l in listeners
-        listener {
-          instance_port     l[:out]
-          instance_protocol "http"
-          lb_port           l[:in]
-          lb_protocol       "http"
-        }
-      end
-    }
+1. **Uses Developer Vocabulary** to build configurations.
+1. **Explicit Configuration**: although there are some short cuts, in general the configuration written is the configuration used. No tricks.
+1. **YAML and JSON Schema** YAML is user friendly configuration language. JSON schema provides strict validation to input. Both these technologies are used to provide useful and quick feedback to developers.
+1. **Extensible Configuration** everyone is different, GPS is a high level tool to build your own extensions.
+1. **Backwards Compatible**: GPS is built to be usable with already defined GeoEngineer repositories.
 
-    @elb    = elb
-    @elb_sg = elb_sg
-  end
+A GPS uses the file location to associate with a project, where the file looks like:
 
-  def template_resources
-    [@elb, @elb_sg]
-  end
-end
+```yml
+<environment_name>:
+  <config_name>:
+    <node_type>:
+      <node_name>:
+        <attributes>:
+```
 
-# Instantiate the template for this project to forward two ports 80 and 8080
-project.from_template("load_balanced_instance", "main_app", {
-  listeners: [{in: 80, out: 3000 }, {in: 8080, out: 4000 }]
-})
+e.g. the project `org/first-project` has the file `./gps/org/first-project.yml` which describes a node `service` named `api` with configuration `staging` in the `development` environment:
+
+```yml
+development:
+  staging:
+    service:
+      api:
+        ports: ["80:80"]
+```
+
+This configuration could be very repetitive. To reduce copy/paste we implement paritals with the key `_`. These  which are merged into a location with `<<:` the `jsonpath` result, e.g.:
+
+```yaml
+_partials:
+  service:
+    api:
+      ports: ["80:80"]
+
+development:
+  staging:
+    <<: ["$._partials"]
```

Comment:
Can you elaborate why you aren't using YAML reference operators to do this? Is it because you intend to implement some custom merge logic eventually? Note that `<<` is an operator in YAML, so you're effectively accomplishing the same thing as a reference from what I can see in this example.

Also note that this has limited uses. You can't modify a single nested attribute in a hash when doing merging this way. Any keys following will clobber ones injected using `<<`.

```yaml
# test.yaml
_partials: &_partials
  service:
    api:
      ports: ["80:80"]

development:
  staging:
    <<: *_partials

production:
  beta:
    <<: *_partials
    service:
      api:
        ports: ["443:8080"]
```

```bash
ruby -e "require 'yaml'; require 'json'; puts JSON.pretty_generate(YAML.load_file('test.yaml'))"
```
```
{
  "_partials": {
    "service": {
      "api": {
        "what": "foo",
        "ports": [
          "80:80"
        ]
      }
    }
  },
  "development": {
    "staging": {
      "service": {
        "api": {
          "what": "foo",
          "ports": [
            "80:80"
          ]
        }
      }
    }
  },
  "production": {
    "beta": {
      "service": {
        "api": {
          "ports": [
            "443:8080"
          ]
        }
      }
    }
  }
}
```